components/org.wso2.carbon.identity.auth.abstractotp/src/main/java/org/wso2/carbon/identity/auth/abstractotp/AbstractOTPAuthenticator.java

```diff
@@ -0,0 +1,1247 @@
+import static org.wso2.carbon.identity.event.IdentityEventConstants.EventProperty.OPERATION_STATUS;
+import static org.wso2.carbon.identity.event.IdentityEventConstants.EventProperty.PROPERTY_FAILED_LOGIN_ATTEMPTS_CLAIM;
+import static org.wso2.carbon.identity.event.IdentityEventConstants.EventProperty.USER_STORE_MANAGER;
+import static org.wso2.carbon.user.core.UserCoreConstants.PRIMARY_DEFAULT_DOMAIN_NAME;
+
+/**
+ * This class contains the implementation of abstract OTP authenticator.
+ */
+public abstract class AbstractOTPAuthenticator extends AbstractApplicationAuthenticator {
+
+    private static final Log log = LogFactory.getLog(AbstractOTPAuthenticator.class);
+
+    @Override
+    public AuthenticatorFlowStatus process(HttpServletRequest request, HttpServletResponse response,
+                                           AuthenticationContext context)
+            throws AuthenticationFailedException, LogoutFailedException {
+
+        AuthenticatorConstants.AuthenticationScenarios scenario = resolveScenario(request, context);
+        switch (scenario) {
+            case LOGOUT:
+                return AuthenticatorFlowStatus.SUCCESS_COMPLETED;
+            case INITIAL_OTP:
+                initiateAuthenticationRequest(request, response, context);
+                return AuthenticatorFlowStatus.INCOMPLETE;
+            default:
+                // Resend OTP and Submit OTP processing will be handled from here.
+                return super.process(request, response, context);
+        }
+    }
+
+    @Override
+    protected void initiateAuthenticationRequest(HttpServletRequest request,
+                                                 HttpServletResponse response, AuthenticationContext context)
+            throws AuthenticationFailedException {
+
+        AuthenticatedUser authenticatedUserFromContext = getAuthenticatedUserFromContext(context);
+        String applicationTenantDomain = context.getTenantDomain();
+
+        /*
+         * If an invalid user has attempted to log in by submitting an OTP code, or if an invalid user has requested
+         * to resend the OTP code, they should be redirected to the OTP login page. This is only valid when OTP is
+         * set as the first step of the authentication flow because in other cases, there will always be a valid user in
+         * the context.
+         */
+        if (isLoginAttemptByInvalidUser(context, authenticatedUserFromContext)) {
+            AuthenticatedUser invalidUser = new AuthenticatedUser();
+            invalidUser.setUserName((String) context.getProperty(INVALID_USERNAME));
+            redirectToOTPLoginPage(invalidUser, applicationTenantDomain, false,
+                    response, request, context);
+            return;
+        }
+        if (isAuthenticatorEnabledForFirstFactor()
+                && authenticatedUserFromContext == null && this instanceof PasswordlessOTPAuthenticator) {
```

Comment:
Can you please explain the meaning of this condition?

Reply:
There can be OTP authenticators that only supports MFA and there could be others that support both first factor and multi factor authentication. By extending the AbstractOTPAuthenticator class, an OTP authenticator by default supports MFA. By implementing the PasswordlessOTPAuthenticator, it adds support for first factor. 

The support for first factor can be enabled and disabled using a different config. For example, it can be defined in `application-authentication.xml`.In that case, even if the authenticator supports first factor in code level, it can be disbaled by the config. That check can be done through `isAuthenticatorEnabledForFirstFactor` method.

`authenticatedUserFromContext == null` is to check if we already have an authenticated user from context. It will be null only in the case where the OTP authenticator is set as the first step of authentication.

We can rearrange the sequence of the above checks in the following order.
1. this instanceof PasswordlessOTPAuthenticator
2. isAuthenticatorEnabledForFirstFactor()
3. authenticatedUserFromContext == null